File: BatteryExtra/BatteryExtra.m
/*
 * Copyright (c) Gábor Librecz <kuglee@gmail.com>
 *
 * This program is free software. It comes without any warranty, to
 * the extent permitted by applicable law. You can redistribute it
 * and/or modify it under the terms of the Do What The Fuck You Want
 * To Public License, Version 2, as published by Sam Hocevar. See
 * http://www.wtfpl.net/ for more details.
 */

#import "ZKSwizzle.h"
#import <AppKit/AppKit.h>

struct powerSourceTypeStruct {
  long long numberOfSources;
  long long numberOfSourcesPresent;
  long long totalNumberOfMinutesLeft;
  long long totalCurrentCapacity;
  long long totalMaxCapacity;
  char isPresent;
  char anySourceCalculating;
  char anySourceUnknownMinutesLeft;
  char lastSourceIsFinishingCharge;
  char anySourceIsPluggedIn;
  char anySourceCharging;
  char allSourcesCharged;
  char anySourceInPoorCondition;
  char anySourceInCheckBatteryCondition;
  char anySourceInFairCondition;
  char anySourceInUnknownCondition;
  double percentageRemaining;
};

@interface NSMenuExtra : NSStatusItem
@end

/// Readds code removed by Apple.
@implementation NSMenuExtra
@end

@interface MyBatteryExtra : NSMenuExtra <NSMenuDelegate>

- (NSString *)_timeStringForMinutes:(unsigned long long)minutes;

@end

@implementation MyBatteryExtra

BOOL *showPerc;
BOOL showTime;
BOOL hideLabelWhenCharged;
BOOL hideIcon;
NSBundle *myBundle;
NSUserDefaults *userDefaults;
extern NSBundle *pluginMainBundle;

- (NSString *)_statusMenuItemTitleWithSourceState:
    (struct powerSourceTypeStruct)state {
  NSString *title = ZKOrig(id, state);

  if (showTime) {
    if (state.allSourcesCharged) {
      if (state.numberOfSourcesPresent == 1)
        title =
            [myBundle localizedStringForKey:@"BAT_CHARGED" value:@""
                                      table:nil];
      else
        title = [myBundle localizedStringForKey:@"BATS_CHARGED"
                                          value:@""
                                          table:nil];
    } else {
      if (state.anySourceIsPluggedIn && !state.anySourceCharging) {
        if (state.numberOfSourcesPresent == 1)
          title = [myBundle localizedStringForKey:@"BAT_NOT_CHARGING"
                                            value:@""
                                            table:nil];
        else
          title = [myBundle localizedStringForKey:@"BATS_NOT_CHARGING"
                                            value:@""
                                            table:nil];
      } else {
        NSString *percentageString = [NSNumberFormatter
            localizedStringFromNumber:
                [NSNumber numberWithFloat:state.percentageRemaining / 100]
                          numberStyle:3];
        NSString *localizedPercentageStringFormat =
            [myBundle localizedStringForKey:@"SINGLE_PERCENT_FORMAT"
                                      value:@""
                                      table:nil];
        NSString *localizedPercentageString =
            [NSString localizedStringWithFormat:localizedPercentageStringFormat,
                                                percentageString];

        if (state.lastSourceIsFinishingCharge) {
          NSString *localizedString =
              [myBundle localizedStringForKey:@"BAT_FINISH_PERC"
                                        value:@""
                                        table:nil];
          title =
              [NSString localizedStringWithFormat:localizedString,
                                                  localizedPercentageString];
        } else {
          NSString *localizedString =
              [myBundle localizedStringForKey:@"PERCENT_FULL"
                                        value:@""
                                        table:nil];
          title =
              [NSString localizedStringWithFormat:localizedString,
                                                  localizedPercentageString];
        }
      }
    }
  } else {
    if (state.allSourcesCharged) {
      if (state.numberOfSourcesPresent == 1)
        title =
            [myBundle localizedStringForKey:@"BAT_CHARGED" value:@""
                                      table:nil];
      else
        title = [myBundle localizedStringForKey:@"BATS_CHARGED"
                                          value:@""
                                          table:nil];
    } else if (state.anySourceCalculating) {
      if (state.anySourceIsPluggedIn)
        title = [myBundle localizedStringForKey:@"CALCULATING_TIME_UNTIL_FULL"
                                          value:@""
                                          table:nil];
      else
        title = [myBundle localizedStringForKey:@"CALCULATING_TIME_REMAINING"
                                          value:@""
                                          table:nil];
    } else if (state.anySourceCharging) {
      NSString *timeStringForMinutes =
          [self _timeStringForMinutes:state.totalNumberOfMinutesLeft];

      if (state.lastSourceIsFinishingCharge)
        title = [NSString localizedStringWithFormat:
                              [myBundle localizedStringForKey:@"BAT_FINISH_TIME"
                                                        value:@""
                                                        table:nil],
                              timeStringForMinutes];
      else {
        if (state.totalNumberOfMinutesLeft > 600)
          title = [NSString
              localizedStringWithFormat:
                  [myBundle localizedStringForKey:@"TIME_UNTIL_FULL_TOO_LONG"
                                            value:@""
                                            table:nil],
                  timeStringForMinutes];
        else
          title =
              [NSString localizedStringWithFormat:
                            [myBundle localizedStringForKey:@"TIME_UNTIL_FULL"
                                                      value:@""
                                                      table:nil],
                            timeStringForMinutes];
      }
    } else if (state.anySourceIsPluggedIn) {
      if (state.numberOfSourcesPresent == 1)
        title = [myBundle localizedStringForKey:@"BAT_NOT_CHARGING"
                                          value:@""
                                          table:nil];
      else
        title = [myBundle localizedStringForKey:@"BATS_NOT_CHARGING"
                                          value:@""
                                          table:nil];

    } else {
      NSString *timeStringForMinutes =
          [self _timeStringForMinutes:state.totalNumberOfMinutesLeft];
      title = [NSString
          localizedStringWithFormat:[myBundle
                                        localizedStringForKey:@"TIME_REMAINING"
                                                        value:@""
                                                        table:nil],
                                    timeStringForMinutes];
    }
  }

  return title;
}

- (NSString *)_menuBarStringWithSourceState:
    (struct powerSourceTypeStruct)state {
  NSString *title = ZKOrig(NSString *, state);

  if (state.allSourcesCharged && hideLabelWhenCharged && !hideIcon) {
    title = nil;
  } else if (showTime && state.numberOfSourcesPresent > 0) {
    if (state.allSourcesCharged) {
      title = [myBundle localizedStringForKey:@"CHARGED" value:@"" table:nil];
    } else if (state.anySourceCalculating)
      title =
          [myBundle localizedStringForKey:@"CALCULATING" value:@"" table:nil];
    else if (state.lastSourceIsFinishingCharge)
      title = [myBundle localizedStringForKey:@"BAT_FINISHING_CHARGE_MENU_BAR"
                                        value:@""
                                        table:nil];
    else if (state.anySourceIsPluggedIn && !state.anySourceCharging)
      title = [myBundle localizedStringForKey:@"BAT_NOT_CHARGING_MENU_BAR"
                                        value:@""
                                        table:nil];
    else if (state.anySourceUnknownMinutesLeft)
      title =
          [myBundle localizedStringForKey:@"UNKNOWN_TIME" value:@"" table:nil];
    else if (state.totalNumberOfMinutesLeft >= 0)
      title = [self _timeStringForMinutes:state.totalNumberOfMinutesLeft];
  }

  return title;
}

- (void)delayedInitBatteryMenu {
  myBundle = ZKHookIvar(self, NSBundle *, "_mybundle");
  userDefaults =
      [[NSUserDefaults alloc] initWithSuiteName:myBundle.bundleIdentifier];

  hideLabelWhenCharged = [userDefaults boolForKey:@"HideLabelWhenCharged"];
  hideIcon = [userDefaults boolForKey:@"HideIcon"];
  showPerc = &ZKHookIvar(self, BOOL, "_showPerc");
  *showPerc = [userDefaults boolForKey:@"ShowPercent"];
  showTime = [userDefaults boolForKey:@"ShowTime"];

  if (showTime && *showPerc) {
    *showPerc = NO;
    [userDefaults setBool:*showPerc forKey:@"ShowPercent"];
  }

  ZKOrig(void);
}

- (NSMenu *)createMenu {
  NSMenu *menu = ZKOrig(NSMenu *);

  NSMenuItem *showMenuItem = [[NSMenuItem alloc]
      initWithTitle:[myBundle localizedStringForKey:@"SHOW" value:@""
                                              table:nil]
             action:nil
      keyEquivalent:@""];
  [menu insertItem:showMenuItem atIndex:menu.itemArray.count - 1];

  NSMenu *showMenu = [[NSMenu alloc] initWithTitle:@""];
  [menu setSubmenu:showMenu forItem:showMenuItem];

  NSMenuItem *showIconOnlyMenuItem = [[NSMenuItem alloc]
      initWithTitle:[myBundle localizedStringForKey:@"NONE" value:@""
                                              table:nil]
             action:@selector(selectShowMenuItem:)
      keyEquivalent:@""];
  showIconOnlyMenuItem.target = self;
  showIconOnlyMenuItem.tag = 3;
  if (!*showPerc && !showTime)
    showIconOnlyMenuItem.state = NSOnState;
  [showMenu addItem:showIconOnlyMenuItem];

  NSMenuItem *showTimeMenuItem = [[NSMenuItem alloc]
      initWithTitle:[myBundle localizedStringForKey:@"DISPLAY_TIME"
                                              value:@""
                                              table:nil]
             action:@selector(selectShowMenuItem:)
      keyEquivalent:@""];
  showTimeMenuItem.target = self;
  showTimeMenuItem.tag = 1;
  if (showTime)
    showTimeMenuItem.state = NSOnState;
  [showMenu addItem:showTimeMenuItem];

  NSMenuItem *showPercentMenuItem =
      ZKHookIvar(self, NSMenuItem *, "_showPercentMenuItem");
  showPercentMenuItem.tag = 2;
  showPercentMenuItem.title =
      [myBundle localizedStringForKey:@"DISPLAY_PERCENT" value:@"" table:nil];
  if (*showPerc)
    showPercentMenuItem.state = NSOnState;
  [menu removeItem:showPercentMenuItem];
  [showMenu addItem:showPercentMenuItem];

  [showMenu addItem:NSMenuItem.separatorItem];

  NSMenuItem *showIcon = [[NSMenuItem alloc]
      initWithTitle:[pluginMainBundle localizedStringForKey:@"SHOW_ICON"
                                                      value:@""
                                                      table:nil]
             action:@selector(selectShowMenuItem:)
      keyEquivalent:@""];
  showIcon.target = self;
  showIcon.tag = 4;

  showIconOnlyMenuItem.state = NSOnState;
  [showMenu addItem:showIcon];

  [menu insertItem:NSMenuItem.separatorItem atIndex:menu.itemArray.count - 1];

  [self performSelector:@selector(updateMenu)];

  return menu;
}

- (void)selectShowMenuItem:(NSMenuItem *)sender {
  switch (sender.tag) {
  case 1:
    showTime = YES;
    *showPerc = NO;
    break;
  case 2:
    showTime = NO;
    *showPerc = YES;
    break;
  case 3:
    showTime = NO;
    *showPerc = NO;
    hideIcon = NO;
    break;
  case 4:
    if (!showTime && !*showPerc)
      *showPerc = YES;

    hideIcon = !hideIcon;
    break;
  }

  [userDefaults setBool:showTime forKey:@"ShowTime"];
  [userDefaults setBool:*showPerc forKey:@"ShowPercent"];
  [userDefaults setBool:hideIcon forKey:@"HideIcon"];

  [self performSelector:@selector(updateMenu)];
}

- (BOOL)validateMenuItem:(NSMenuItem *)item {
  switch (item.tag) {
  case 1:
    item.state = showTime ? NSOnState : NSOffState;
    break;
  case 2:
    item.state = *showPerc ? NSOnState : NSOffState;
    break;
  case 3:
    item.state =
        (!showTime && !*showPerc && !hideIcon) ? NSOnState : NSOffState;
    break;
  case 4:
    item.state = !hideIcon;
    break;
  default:
    break;
  }

  return YES;
}

@end

File: BatteryExtra/BatteryExtra.swift
/*
 * Copyright (c) Gábor Librecz <kuglee@gmail.com>
 *
 * This program is free software. It comes without any warranty, to
 * the extent permitted by applicable law. You can redistribute it
 * and/or modify it under the terms of the Do What The Fuck You Want
 * To Public License, Version 2, as published by Sam Hocevar. See
 * http://www.wtfpl.net/ for more details.
 */

import AppKit

fileprivate var showTime: Bool!
fileprivate var hideLabelWhenCharged: Bool!
var hideIcon: Bool!
var userDefaults: UserDefaults!

typealias BatteryExtra = NSStatusItem

/// Readds code removed by Apple.
extension BatteryExtra {
  var showPercent: Bool {
    get {
      return value(forKey: "_showPerc") as! Bool
    }
    set {
      setValue(newValue, forKey: "_showPerc")
    }
  }

  var mybundle: Bundle {
    get {
      return value(forKey: "_mybundle") as! Bundle
    }
    set {
      setValue(newValue, forKey: "_mybundle")
    }
  }

  @objc func swizzled_timeStringForMinutes(_ minutes: CUnsignedLongLong) -> NSString {
    return swizzled_timeStringForMinutes(minutes)
  }

  @objc func swizzledUpdateMenu() {
    swizzledUpdateMenu()
  }

  @objc func swizzled_statusMenuItemTitleWithSourceState(_ sourceState: powerSourceTypeStruct) -> NSString {
    var title = swizzled_statusMenuItemTitleWithSourceState(sourceState) as String

    if showTime {
      if sourceState.allSourcesCharged > 0 {
        if sourceState.numberOfSourcesPresent == 1 {
          title = mybundle.localizedString(forKey: "BAT_CHARGED", value: "", table: nil)
        } else {
          title = mybundle.localizedString(forKey: "BATS_CHARGED", value: "", table: nil)
        }
      } else {
        if sourceState.anySourceIsPluggedIn > 0, !(sourceState.anySourceCharging > 0) {
          if sourceState.numberOfSourcesPresent == 1 {
            title = mybundle.localizedString(forKey: "BAT_NOT_CHARGING", value: "", table: nil)
          } else {
            title = mybundle.localizedString(forKey: "BATS_NOT_CHARGING", value: "", table: nil)
          }
        } else {
          let percentageString = NumberFormatter.localizedString(from: NSNumber(floatLiteral: sourceState.percentageRemaining / 100), number: NumberFormatter.Style.percent)

          let localizedPercentageStringFormat = mybundle.localizedString(forKey: "SINGLE_PERCENT_FORMAT", value: "", table: nil)

          let localizedPercentageString = String.localizedStringWithFormat(localizedPercentageStringFormat, percentageString)

          if sourceState.lastSourceIsFinishingCharge > 0 {
            let localizedString = mybundle.localizedString(forKey: "BAT_FINISH_PERC", value: "", table: nil)
            title = String.localizedStringWithFormat(localizedString, localizedPercentageString)
          } else {
            let localizedString = mybundle.localizedString(forKey: "PERCENT_FULL", value: "", table: nil)
            title = String.localizedStringWithFormat(localizedString, localizedPercentageString)
          }
        }
      }
    } else {
      if sourceState.allSourcesCharged > 0 {
        if sourceState.numberOfSourcesPresent == 1 {
          title = mybundle.localizedString(forKey: "BAT_CHARGED", value: "", table: nil)
        } else {
          title = mybundle.localizedString(forKey: "BATS_CHARGED", value: "", table: nil)
        }
      } else if sourceState.anySourceCalculating > 0 {
        if sourceState.anySourceIsPluggedIn > 0 {
          title = mybundle.localizedString(forKey: "CALCULATING_TIME_UNTIL_FULL", value: "", table: nil)
        } else {
          title = mybundle.localizedString(forKey: "CALCULATING_TIME_REMAINING", value: "", table: nil)
        }
      } else if sourceState.anySourceCharging > 0 {
        let timeStringForMinutes = swizzled_timeStringForMinutes(CUnsignedLongLong(sourceState.totalNumberOfMinutesLeft))

        if sourceState.lastSourceIsFinishingCharge > 0 {
          title = String.localizedStringWithFormat(mybundle.localizedString(forKey: "BAT_FINISH_TIME", value: "", table: nil), timeStringForMinutes)
        } else {
          if sourceState.totalNumberOfMinutesLeft > 600 {
            title = String.localizedStringWithFormat(mybundle.localizedString(forKey: "TIME_UNTIL_FULL_TOO_LONG", value: "", table: nil), timeStringForMinutes)
          } else {
            title = String.localizedStringWithFormat(mybundle.localizedString(forKey: "TIME_UNTIL_FULL", value: "", table: nil), timeStringForMinutes)
          }
        }
      } else if sourceState.anySourceIsPluggedIn > 0 {
        if sourceState.numberOfSourcesPresent == 1 {
          title = mybundle.localizedString(forKey: "BAT_NOT_CHARGING", value: "", table: nil)
        } else {
          title = mybundle.localizedString(forKey: "BATS_NOT_CHARGING", value: "", table: nil)
        }
      } else {
        let timeStringForMinutes = swizzled_timeStringForMinutes(CUnsignedLongLong(sourceState.totalNumberOfMinutesLeft))
        title = String.localizedStringWithFormat(mybundle.localizedString(forKey: "TIME_REMAINING", value: "", table: nil), timeStringForMinutes)
      }
    }

    return title as NSString
  }

  @objc func swizzled_menuBarStringWithSourceState(_ sourceState: powerSourceTypeStruct) -> NSString {
    var title = swizzled_menuBarStringWithSourceState(sourceState) as String

    if sourceState.allSourcesCharged > 0, hideLabelWhenCharged, !hideIcon {
      title = ""
    } else
    if showTime, sourceState.numberOfSourcesPresent > 0 {
      if sourceState.allSourcesCharged > 0 {
        title = mybundle.localizedString(forKey: "CHARGED", value: "", table: nil)
      } else if sourceState.anySourceCalculating > 0 {
        title = mybundle.localizedString(forKey: "CALCULATING", value: "", table: nil)
      } else if sourceState.lastSourceIsFinishingCharge > 0 {
        title = mybundle.localizedString(forKey: "BAT_FINISHING_CHARGE_MENU_BAR", value: "", table: nil)
      } else if sourceState.anySourceIsPluggedIn > 0, !(sourceState.anySourceCharging > 0) {
        title = mybundle.localizedString(forKey: "BAT_NOT_CHARGING_MENU_BAR", value: "", table: nil)
      } else if sourceState.anySourceUnknownMinutesLeft > 0 {
        title = mybundle.localizedString(forKey: "UNKNOWN_TIME", value: "", table: nil)
      } else if sourceState.totalNumberOfMinutesLeft >= 0 {
        title = swizzled_timeStringForMinutes(CUnsignedLongLong(sourceState.totalNumberOfMinutesLeft)) as String
      }
    }

    return title as NSString
  }

  @objc func swizzledDelayedInitBatteryMenu() {
    userDefaults = UserDefaults(suiteName: mybundle.bundleIdentifier)!

    hideLabelWhenCharged = userDefaults.bool(forKey: "HideLabelWhenCharged")
    hideIcon = userDefaults.bool(forKey: "HideIcon")
    showPercent = userDefaults.bool(forKey: "ShowPercent")
    showTime = userDefaults.bool(forKey: "ShowTime")
    hideIcon = userDefaults.bool(forKey: "HideIcon")

    if showTime, showPercent {
      showPercent = false
      userDefaults.set(showPercent, forKey: "ShowPercent")
    }

    swizzledDelayedInitBatteryMenu()
  }

  @objc func swizzledCreateMenu() -> NSMenu {
    let menu = swizzledCreateMenu()

    let showMenuItem = NSMenuItem(title: mybundle.localizedString(forKey: "SHOW", value: "", table: nil), action: nil, keyEquivalent: "")
    menu.insertItem(showMenuItem, at: menu.items.count - 1)

    let showMenu = NSMenu(title: "")
    menu.setSubmenu(showMenu, for: showMenuItem)

    let showIconOnlyMenuItem = NSMenuItem(title: mybundle.localizedString(forKey: "NONE", value: "", table: nil), action: NSSelectorFromString("selectShowMenuItem:"), keyEquivalent: "")
    showIconOnlyMenuItem.target = self
    showIconOnlyMenuItem.tag = 3
    if !showPercent, !showTime {
      showIconOnlyMenuItem.state = NSControl.StateValue.on
    }
    showMenu.addItem(showIconOnlyMenuItem)

    let showTimeMenuItem = NSMenuItem(title: mybundle.localizedString(forKey: "DISPLAY_TIME", value: "", table: nil), action: NSSelectorFromString("selectShowMenuItem:"), keyEquivalent: "")
    showTimeMenuItem.target = self
    showTimeMenuItem.tag = 1
    if showTime {
      showTimeMenuItem.state = NSControl.StateValue.on
    }
    showMenu.addItem(showTimeMenuItem)

    let showPercentMenuItem = value(forKey: "_showPercentMenuItem") as! NSMenuItem
    showPercentMenuItem.title = mybundle.localizedString(forKey: "DISPLAY_PERCENT", value: "", table: nil)
    showPercentMenuItem.tag = 2
    if showPercent {
      showPercentMenuItem.state = NSControl.StateValue.on
    }

    menu.removeItem(showPercentMenuItem)
    showMenu.addItem(showPercentMenuItem)

    showMenu.addItem(NSMenuItem.separator())

    let showIcon = NSMenuItem(title: pluginMainBundle.localizedString(forKey: "SHOW_ICON", value: "", table: nil), action: NSSelectorFromString("selectShowMenuItem:"), keyEquivalent: "")
    showIcon.target = self
    showIcon.tag = 4
    showIcon.state = NSControl.StateValue.on
    showMenu.addItem(showIcon)

    menu.insertItem(NSMenuItem.separator(), at: menu.items.count - 1)

    swizzledUpdateMenu()

    return menu
  }

  @objc func swizzledSelectShowMenuItem(_ menuItem: NSMenuItem) {
    switch menuItem.tag {
    case 1:
      showTime = true
      showPercent = false
    case 2:
      showTime = false
      showPercent = true
    case 3:
      showTime = false
      showPercent = false
      hideIcon = false
    case 4:
      if !showTime, !showPercent {
        showPercent = true
      }

      hideIcon = !hideIcon
    default:
      break
    }

    userDefaults.set(showTime, forKey: "ShowTime")
    userDefaults.set(showPercent, forKey: "ShowPercent")
    userDefaults.set(hideIcon, forKey: "HideIcon")

    swizzledUpdateMenu()
  }

  @objc func swizzledValidateMenuItem(_ menuItem: NSMenuItem) -> Bool {
    switch menuItem.tag {
    case 1:
      menuItem.state = showTime ? NSControl.StateValue.on : NSControl.StateValue.off
    case 2:
      menuItem.state = showPercent ? NSControl.StateValue.on : NSControl.StateValue.off
    case 3:
      menuItem.state = (!showTime && !showPercent && !hideIcon) ? NSControl.StateValue.on : NSControl.StateValue.off
    case 4:
      menuItem.state = !hideIcon ? NSControl.StateValue.on : NSControl.StateValue.off
    default:
      break
    }

    return true
  }
}
